builtin/providers/cloudstack/resource_cloudstack_instance.go

```diff
@@ -186,6 +195,7 @@ func resourceCloudStackInstanceRead(d *schema.ResourceData, meta interface{}) er
 	d.Set("display_name", vm.Displayname)
 	d.Set("ipaddress", vm.Nic[0].Ipaddress)
 	d.Set("zone", vm.Zonename)
+	//NB cloudstack sometimes sends back the wrong keypair name, so dont update it
```

Comment:
Thanks for that! I didn't have time to chase it down, but I guessed it would be to do with the fact I was testing with multiple keypairs registered at Cloudstack that in fact were pointing to the same underlying key values